// @jupyterlab/debugger/lib/variables/index.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
import { Panel } from '@lumino/widgets';
import { VariablesBodyTable } from './table';
import { VariablesHeader } from './header';
import { ToolbarButton } from '@jupyterlab/apputils';
import { VariablesBodyTree } from './tree';
/**
 * A Panel to show a variable explorer.
 */
export class Variables extends Panel {
    /**
     * Instantiate a new Variables Panel.
     * @param options The instantiation options for a Variables Panel.
     */
    constructor(options) {
        super();
        const { model, service, commands } = options;
        this._header = new VariablesHeader();
        this._tree = new VariablesBodyTree({ model, service });
        this._table = new VariablesBodyTable({ model, commands });
        this._table.hide();
        const onClick = () => {
            if (this._table.isHidden) {
                this._tree.hide();
                this._table.show();
                this.node.setAttribute('data-jp-table', 'true');
            }
            else {
                this._tree.show();
                this._table.hide();
                this.node.removeAttribute('data-jp-table');
            }
            this.update();
        };
        this._header.toolbar.addItem('view-VariableSwitch', new ToolbarButton({
            iconClass: 'jp-ToggleSwitch',
            onClick,
            tooltip: 'Table / Tree View'
        }));
        this.addWidget(this._header);
        this.addWidget(this._tree);
        this.addWidget(this._table);
        this.addClass('jp-DebuggerVariables');
    }
    /**
     * Set the variable filter for both the tree and table views.
     */
    set filter(filter) {
        this._tree.filter = filter;
        this._table.filter = filter;
    }
    /**
     * A message handler invoked on a `'resize'` message.
     */
    onResize(msg) {
        super.onResize(msg);
        this._resizeBody(msg);
    }
    /**
     * Resize the body.
     * @param msg The resize message.
     */
    _resizeBody(msg) {
        const height = msg.height - this._header.node.offsetHeight;
        this._table.node.style.height = `${height}px`;
        this._tree.node.style.height = `${height}px`;
    }
}
/**
 * Convert a variable to a primitive type.
 * @param variable The variable.
 */
export const convertType = (variable) => {
    const { type, value } = variable;
    switch (type) {
        case 'int':
            return parseInt(value, 10);
        case 'float':
            return parseFloat(value);
        case 'bool':
            return value;
        case 'str':
            return value.slice(1, value.length - 1);
        default:
            return type;
    }
};
